File: test/features/knowledge/domain/knowledge_filters_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lexi_core/features/dictionary/domain/entities/language.dart';
import 'package:lexi_core/features/vocabulary/domain/entities/cefr_level.dart';
import 'package:lexi_core/features/knowledge/domain/entities/knowledge_note.dart';
import 'package:lexi_core/features/knowledge/domain/knowledge_filters.dart';

KnowledgeNote _n(String id, {
  String title = '', String explanation = '', String group = 'en_other',
  CEFRLevel? level, List<String> tags = const [],
}) {
  final t = DateTime.utc(2026, 1, 1);
  return KnowledgeNote(
    id: id, title: title, summary: '', explanation: explanation, groupId: group,
    tags: tags, cefrLevel: level, targetLanguage: Language.english,
    origin: KnowledgeNoteOrigin.manual, createdAt: t, updatedAt: t,
  );
}

void main() {
  final notes = [
    _n('a', title: 'Câu điều kiện loại 2', group: 'en_conditionals', level: CEFRLevel.b1, tags: ['toeic']),
    _n('b', title: 'Thì hiện tại đơn', group: 'en_tenses', level: CEFRLevel.a1),
    _n('c', explanation: 'nói về **điều kiện** trong quá khứ', group: 'en_conditionals', level: CEFRLevel.b2),
  ];

  test('query matches folded across fields', () {
    final r = applyKnowledgeFilter(notes,
        const KnowledgeFilter(query: 'dieu kien'));
    expect(r.map((n) => n.id), containsAll(['a', 'c']));
  });

  test('group + level filters AND across, OR within', () {
    final r = applyKnowledgeFilter(notes, const KnowledgeFilter(
      groupIds: {'en_conditionals'}, levels: {CEFRLevel.b1},
    ));
    expect(r.map((n) => n.id), ['a']);
  });

  test('tag filter', () {
    final r = applyKnowledgeFilter(notes, const KnowledgeFilter(tags: {'toeic'}));
    expect(r.map((n) => n.id), ['a']);
  });

  test('group counts and all tags', () {
    expect(knowledgeGroupCounts(notes)['en_conditionals'], 2);
    expect(knowledgeAllTags(notes), ['toeic']);
  });
}

File: apps/web/src/lib/knowledgeFilters.test.ts
import { describe, expect, it } from "vitest";
import { applyKnowledgeFilter, knowledgeGroupCounts, knowledgeAllTags, type KnowledgeFilter } from "./knowledgeFilters";
import type { KnowledgeNote } from "./knowledgeNotes";

const n = (over: Partial<KnowledgeNote>): KnowledgeNote => ({
  id: "x", title: "", summary: "", explanation: "", patterns: [], examples: [], pitfalls: [],
  groupId: "en_other", tags: [], cefrLevel: null, targetLanguage: "english", source: "manual",
  sourcePrompt: null, createdAt: "2026-01-01T00:00:00.000Z", updatedAt: "2026-01-01T00:00:00.000Z", ...over,
});
const empty = (o: Partial<KnowledgeFilter> = {}): KnowledgeFilter =>
  ({ query: "", groupIds: new Set(), levels: new Set(), tags: new Set(), ...o });

const notes = [
  n({ id: "a", title: "Câu điều kiện loại 2", groupId: "en_conditionals", cefrLevel: "b1", tags: ["toeic"] }),
  n({ id: "b", title: "Thì hiện tại đơn", groupId: "en_tenses", cefrLevel: "a1" }),
  n({ id: "c", explanation: "nói về **điều kiện**", groupId: "en_conditionals", cefrLevel: "b2" }),
];

describe("applyKnowledgeFilter", () => {
  it("folded query across fields", () => {
    expect(applyKnowledgeFilter(notes, empty({ query: "dieu kien" })).map((x) => x.id).sort()).toEqual(["a", "c"]);
  });
  it("group + level AND across, OR within", () => {
    expect(applyKnowledgeFilter(notes, empty({ groupIds: new Set(["en_conditionals"]), levels: new Set(["b1"]) }))
      .map((x) => x.id)).toEqual(["a"]);
  });
  it("tag filter", () => {
    expect(applyKnowledgeFilter(notes, empty({ tags: new Set(["toeic"]) })).map((x) => x.id)).toEqual(["a"]);
  });
});
describe("counts + tags", () => {
  it("group counts and sorted unique tags", () => {
    expect(knowledgeGroupCounts(notes).en_conditionals).toBe(2);
    expect(knowledgeAllTags(notes)).toEqual(["toeic"]);
  });
});
